README.md
# List comprehensions in R

[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/comprehenr)](https://cran.r-project.org/package=comprehenr)
[![](https://cranlogs.r-pkg.org/badges/comprehenr)](https://cran.rstudio.com/web/packages/comprehenr/index.html)
[![](https://cranlogs.r-pkg.org/badges/grand-total/comprehenr)](https://cran.rstudio.com/web/packages/comprehenr/index.html)
[![Coverage Status](https://img.shields.io/codecov/c/github/gdemin/comprehenr/master.svg)](https://codecov.io/github/gdemin/comprehenr?branch=master)

Package provides Python-style list comprehensions for R. List comprehension
expressions use usual loops (`for`, `while` and `repeat`) and usual `if` as
list producers. Syntax is very similar to Python. The difference is that
returned value should be at the end of the loop body.

There are three main functions:

- `to_list` converts usual R loops expressions to list producers. Expression should be started with `for`, `while` or `repeat`. You can iterate over multiple lists if you provide several loop variables in backticks. See examples.
- `to_vec` is the same as `to_list` but return vector. See examples.
- `alter` return the same type as its argument but with modified elements. It is useful for altering existing data.frames or lists. See examples.

Rather unpractical example - squares of even numbers:
```R
library(comprehenr)
to_vec(for(i in 1:10) if(i %% 2==0) i*i)
```
Pythagorean triples:
```R
to_list(for (x in 1:20) for (y in x:20) for (z in y:20) if (x^2 + y^2 == z^2) c(x, y, z))
```
More examples:
```R
colours = c("red", "green", "yellow", "blue")
things = c("house", "car", "tree")
to_vec(for(x in colours) for(y in things) paste(x, y))

# prime numbers
noprimes = to_vec(for (i in 2:7) for (j in seq(i*2, 99, i)) j)
primes = to_vec(for (x in 2:99) if(!x %in% noprimes) x)
primes
```
You can iterate over multiple lists if you provide several loop variables in backticks:
```R
to_vec(for(`i, j` in numerate(letters)) if(i %% 2==0) paste(i, j))

set.seed(123)
rand_sequence = runif(20)
# gives only locally increasing values
to_vec(for(`i, j` in lag_list(rand_sequence)) if(j>i) j)

```

`alter` examples:
```R
data(iris)
# scale numeric variables
res = alter(for(i in iris) if(is.numeric(i)) scale(i))
str(res)

# convert factors to characters
res = alter(for(i in iris) if(is.factor(i)) as.character(i))
str(res)

# drop factors
res = alter(for(i in iris) if(is.factor(i)) exclude())
str(res)

# 'data' argument example
# specify which columns to map with a numeric vector of positions:
res = alter(
    for(`i, value` in numerate(mtcars)) if(i %in% c(1, 4, 5)) as.character(value),
    data = mtcars
)
str(res)

# or with a vector of names:
res = alter(
    for(`name, value` in mark(mtcars)) if(name %in% c("cyl", "am")) as.character(value),
    data = mtcars
)
str(res)
```


NEWS
0.6.10 (31-01-2021)
================
* add new function 'to_df' which combine result of list comprehension to data.frame (issue #2)

0.6.9 (18-01-2021)
================
* fix bug with nested 'to_list'/'to_vec'/'alter' (issue #1)

0.6.8 (22-07-2020)
================
* maintainance release
* convert documentation to markdown
* add rmarkdown to SUGGESTS
* move tests from testthat to tinytest

0.6.7 (17.06.2019)
================
* urgent fix for bug with incorrect counter in 'alter' when used with 'exclude'
* change behaviour 'lag_list':  x -> list(x[[i-1]], x[[i]])

0.6.6 (07.06.2019)
================
* add 'exclude' function for dropping elements from data.frame/list in 'alter'

0.6.5 (09.04.2019)
================
* add 'alter' function for conditional lists/data.frames modification

0.6.0 (18.03.2019)
===============
* add iterations over multiple lists: to_vec(for(`i, j` in numerate(letters)) if(i %% 2==0) paste(i, j))
* add function 'lag_list':  x -> list(x[i-1], x[i])

0.5.4 (06.03.2019)
================
* Initial release


tests/testthat.R
testthat::test_check("comprehenr")


tests/tinytest.R
if(capabilities('long.double') && requireNamespace("tinytest", quietly=TRUE)){
    library(tinytest)
    test_package("comprehenr")
}


inst/tinytest/test_to_list.R
context = function(x) cat("\nContext:", x, "\n")
context("to_list")

expect_identical(
    to_list(for(i in 1:10) if(i %% 2 !=0) i),
    list(1L,3L,5L,7L,9L)
)

expect_identical(
    to_list(for(i in 1:10) if(i %% 2 !=0) i else exclude()),
    list(1L,3L,5L,7L,9L)
)

expect_identical(
    to_list(for(i in 1:10) if(i>1000) i),
    list()
)

expect_equal(
    to_list(for(i in 1:3) for(j in 1:3) if(i==j) c(i,j)),
    list(c(1,1), c(2,2), c(3,3))
)

expect_equal(
    to_list(for(i in 1:3) if(i==2) for(k in 1:2) k*2 else for(l in 3:4) l*3),
    list(9, 12, 2, 4, 9, 12)
)

expect_equal(
    to_list(for (x in 1:29) for (y in x:29) for (z in y:29) if (x^2 + y^2 == z^2) c(x, y, z)),
    list(c(3, 4, 5),
         c(5, 12, 13),
         c(6, 8, 10),
         c(7, 24, 25),
         c(8, 15, 17),
         c(9, 12, 15),
         c(10, 24, 26),
         c(12, 16, 20),
         c(15, 20, 25),
         c(20, 21, 29)
    )
)
expect_false(exists(".___res"))
expect_false(exists(".___counter"))
expect_false(exists(".___curr"))

expect_error(to_list(for(i in 1:10) i*brrrr))
expect_false(exists(".___res"))
expect_false(exists(".___counter"))
expect_false(exists(".___curr"))


expect_error(to_list(1))
expect_false(exists(".___res"))
expect_false(exists(".___counter"))
expect_false(exists(".___curr"))
a = 11
expect_error(to_list(a + 2))

context("to_list scoping")

a_global = 3
expect_identical(
    to_list(for(i in 1:10) if(i %% 2 !=0) i*a_global),
    list(1L*a_global,3L*a_global,5L*a_global,7L*a_global,9L*a_global)
)

my_fun = function(curr){
    to_list(for(i in curr) if(i %% 2 !=0) i*a_global)
}
expect_identical(
    my_fun(1:10),
    list(1L*a_global,3L*a_global,5L*a_global,7L*a_global,9L*a_global)
)

my_fun = function(curr){
    a_internal = 4
    to_list(for(i in curr) if(i %% 2 !=0) i*a_internal)
}

expect_identical(
    my_fun(1:10),
    list(1L*4,3L*4,5L*4,7L*4,9L*4)
)

data(BOD)
res = to_list(for(j in 1:2) to_list(for(i in 1:2) BOD[i, j]))
expect_identical(
    res,
    list(list(1, 2), list(8.3, 10.3))
)

res = to_vec(for(j in 1:2) sum(to_vec(for(i in 1:6) BOD[i, j])))
expect_identical(
    res,
    c(sum(BOD[[1]]), sum(BOD[[2]]))
)

context("to_vec")

expect_identical(
    to_vec(for(i in 1:10) if(i %% 2 !=0) i),
    c(1L,3L,5L,7L,9L)
)

expect_identical(
    to_vec(for(i in 1:10) if(i>1000) i),
    c()
)


expect_equal(
    to_vec(for(i in 1:10) i*i),
    (1:10)^2
)

a = 11
i = 0
expect_equal(
    to_vec(while(i<a){
        i = i + 2
        i*i

    }), c(2, 4, 6, 8, 10, 12)^2)


i = 0
expect_equal(
    to_vec(repeat{
        i = i + 2
        if(i>12) break
        i*i
    }), c(2, 4, 6, 8, 10, 12)^2)


data(iris)

expect_equal(
    to_vec(for(i in colnames(iris)) if(is.numeric(iris[[i]])) setNames(mean(iris[[i]]), i) ),
    unname(colMeans(iris[,-5]))
)
expect_equal(
    to_vec(for(i in colnames(iris)) if(is.numeric(iris[[i]])) setNames(mean(iris[[i]]), i) , use.names = TRUE),
    colMeans(iris[,-5])
)


context("`i, j`")

res = to_vec(for(`i, j` in zip_lists(letters, LETTERS)) paste(i, j))
expect_identical(res, paste(letters, LETTERS))


res = to_vec(for(`i, j` in numerate(letters)) if(i %% 2==0) paste(i, j))
index = seq_along(letters)
expect_identical(res, paste(index[index %% 2==0], letters[index %% 2==0]))


res = to_vec(for(`i, j` in numerate(letters)) if(i %% 2==0) for(`k, z` in numerate(LETTERS)) paste(i, j, k, z))
index = seq_along(letters)
true_res = expand.grid(paste(seq_along(letters), LETTERS), paste(index[index %% 2==0], letters[index %% 2==0]))
true_res = paste(true_res[[2]], true_res[[1]])
expect_identical(res, true_res)


context("alter")

data(iris)
iris2 = alter(for(i in iris) if(is.numeric(i)) scale(i))
res_iris = iris
res_iris[,-5] = lapply(iris[,-5], scale)
expect_equal(iris2, res_iris)

data(iris)
iris2 = alter(for(i in iris) if(is.factor(i)) exclude())
res_iris = iris[,-5]
expect_equal(iris2, res_iris)

iris2 = alter(for(`name, value` in mark(iris)) if(endsWith(name, "Width")) scale(value), data = iris)

res_iris = iris
res_iris[,c("Sepal.Width", "Petal.Width")] = lapply(iris[,c("Sepal.Width", "Petal.Width")], scale)
expect_equal(iris2, res_iris)

expect_error(alter(1))
# library(data.table)
# dt_iris = as.data.table(iris)
# dt_iris2 = alter(for(i in dt_iris) if(is.numeric(i)) scale(i))

data(mtcars)
mtcars2= alter(for(i in mtcars) if(length(unique(i))==2) exclude()) # drop am, vs
res_mtcars = mtcars
res_mtcars[, c("am", "vs")]  = NULL
expect_equal(mtcars2, res_mtcars)

data(mtcars)
mtcars2= alter(for(`i,j` in numerate(mtcars)) if(i %in% c(8, 9)) exclude()) # drop am, vs
res_mtcars = mtcars
res_mtcars[, c("am", "vs")]  = NULL
expect_equal(mtcars2, res_mtcars)


data(mtcars)
mtcars2= alter(for(`i,j` in enumerate(mtcars)) if(i %in% c(8, 9)) exclude()) # drop am, vs
res_mtcars = mtcars
res_mtcars[, c("am", "vs")]  = NULL
expect_equal(mtcars2, res_mtcars)

data(mtcars)
mtcars2= alter(for(`i,j` in mark(mtcars)) if(i %in% c("am", "vs")) exclude()) # drop am, vs
res_mtcars = mtcars
res_mtcars[, c("am", "vs")]  = NULL
expect_equal(mtcars2, res_mtcars)

data(mtcars)
mtcars2= alter(for(`i,j` in lag_list(mtcars)) if(length(unique(i))==2) exclude()) # drop am, vs
res_mtcars = mtcars
res_mtcars[, c("am", "vs")]  = NULL
expect_equal(mtcars2, res_mtcars)

context("to_df/to_dfc")

data("mtcars")

expect_identical(
    to_df(for(each in list(1:2, 2:3, 3:4)) each),
    structure(list(X1L = c(1L, NA, NA), X2L = c(2L, 2L, NA), X3L = c(NA,
                                                                     3L, 3L), X4L = c(NA, NA, 4L)), row.names = c(NA, -3L), class = "data.frame")
)

expect_error(
    to_df(for(each in list(1:2, 2:3, 3:4)) each, fill = FALSE),
)

expect_equal(
    to_df(for(`.name, .x` in mark(mtcars)) list(mean = mean(.x), sd = sd(.x), var = .name)),
    data.frame(mean = colMeans(mtcars), sd = sapply(mtcars, sd), var = names(mtcars), row.names = seq_along(mtcars))
)


expect_identical(
    to_df(for(`.index, .x` in enumerate(mtcars)) list(mean = mean(.x), sd = sd(.x), item_id = .index)),
    data.frame(mean = colMeans(mtcars), sd = sapply(mtcars, sd), item_id = seq_along(mtcars), row.names = seq_along(mtcars))
)






R/numerate.R
#' Auxiliary functions for working with lists
#'
#' - `numerate` returns list of lists. Each list consists of two elements:
#' sequential number of element and element. Reverse operation -
#' `unumerate`.
#' - `mark` returns list of lists. Each list consists of two elements: name of element and element. Reverse operation -
#' `unmark`.
#' - `zip_lists` combines lists side-by-sidy. Reverse operation - `unzip_list`.
#' - `unzip_list` is similair to matrix transposition but for list of lists.
#' - `lag_list` converts argument to list of arguments with previous values: `x` -> `list(x[i-1], x[i])`.
#'
#' @param x list, vector or list of lists
#' @param item numeric number of list in which stored values
#' @param ... lists which will be zipped
#'
#' @return list or list of lists
#' @export
#'
#' @examples
#' cities = c('Chicago', 'Detroit', 'Atlanta')
#' airports = c('ORD', 'DTW', 'ATL')
#' pairs = zip_lists(cities, airports)
#'
#' str(pairs)
#' str(unzip_list(pairs))
#'
#' str(numerate(cities))
#'
#' named_list = c('Chicago' = 'ORD', 'Detroit' = 'DTW', 'Atlanta' = 'ATL')
#' str(mark(named_list))
#'
#' set.seed(123)
#' rand_sequence = runif(20)
#' # gives only locally increasing values
#' to_vec(for(`i, j` in lag_list(rand_sequence)) if(j>i) j)
numerate = function(x){
    unzip_list(list(seq_along(x), x))
}

#' @export
#' @rdname numerate
enumerate = numerate

#' @export
#' @rdname numerate
unnumerate = function(x, item = 2){
    unzip_list(x)[[item]]
}

#' @export
#' @rdname numerate
mark = function(x){
    all_names = names(x)
    if(is.null(all_names)) all_names = rep("", length(x))
    unzip_list(list(all_names, x))

}


#' @export
#' @rdname numerate
unmark = function(x, item = 2){
    res = unzip_list(x)
    set_names(
        res[[item]],
        unlist(res[[1]], use.names = FALSE)
        )
}


#' @export
#' @rdname numerate
unzip_list = function(x){
    all_lengths = unique(lengths(x))
    if(length(x)==0) return(x)
    (length(all_lengths) == 1) ||
        stop("each element of the argument 'x' should have the same length.")

    global_names = names(x)
    local_names = lapply(x, names)
    get_names = function(i){
        res = lapply(local_names, "[[", i)
        all_lengths = lengths(res)
        if(all(all_lengths==0)) {
            return(global_names)
        }
        res[all_lengths==0] = ""
        unlist(res, use.names = FALSE)

    }
    lapply(seq_along(x[[1]]), function(i) set_names(lapply(x, `[[`, i), get_names(i)))
}

#' @export
#' @rdname numerate
zip_lists = function(...){
    unzip_list(list(...))
}

#' @export
#' @rdname numerate
lag_list = function(x){
    lapply(seq_along(x)[-1], function(i) list(x[[i-1]], x[[i]]))
}


set_names = function (object = nm, nm) {
    names(object) = nm
    object
}


R/enumerate.R


numerate = function(x){
    lapply(seq_along(x), function(i) list(i, x[[i]]))
}

enumerate = numerate

mark = function(x){
    all_names = names(x)
    if(is.null(all_names)){
        lapply(seq_along(x), function(i) list("", x[[i]]))
    } else {
        lapply(seq_along(x), function(i) list(all_names[i], x[[i]]))
    }

}

unmark = function(x, item = 2){
    all_names = unlist(unnumerate(x, item = 1), use.names = FALSE)
    all_values = unnumerate(x, item = item)
    setNames(all_values, all_names)
}


unnumerate = function(x, item = 2){
    all_lengths = unique(lengths(x))
    if(length(all_lengths)!=1 || any(all_lengths<item)){
        stop(paste0("'denumerate': each element of the argument should have the same length greater or equal to ", item))
    }
    lapply(x, `[[`, item)
}


transpose = function(x){
    all_lengths = unique(lengths(x))
    if(length(x)==0) return(x)
    if(length(all_lengths)!=1){
        stop("'transpose': each element of the argument should have the same length.")
    }
    all_names = lapply(x, names)
    get_names = function(i){
        res = lapply(all_names, "[[", i)
        all_lengths = lengths(res)
        if(all(all_lengths==0)) return(NULL)
        res[all_lengths==0] = ""
        unlist(res, use.names = FALSE)

    }
    lapply(seq_along(x[[1]]), function(i) setNames(lapply(x, `[[`, i), get_names(i)))
}


zip_lists = function(...){
    transpose(list(...))
}


vignettes/Introduction.R
## ----setup, include = FALSE----------------------------------------------
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

## ------------------------------------------------------------------------
library(comprehenr)
to_vec(for(i in 1:10) if(i %% 2==0) i*i)

## ------------------------------------------------------------------------
to_list(for (x in 1:20) for (y in x:20) for (z in y:20) if (x^2 + y^2 == z^2) c(x, y, z))

## ------------------------------------------------------------------------
colours = c("red", "green", "yellow", "blue")
things = c("house", "car", "tree")
to_vec(for(x in colours) for(y in things) paste(x, y))

# prime numbers
noprimes = to_vec(for (i in 2:7) for (j in seq(i*2, 99, i)) j)
primes = to_vec(for (x in 2:99) if(!x %in% noprimes) x)
primes

## ------------------------------------------------------------------------
to_vec(for(`i, j` in numerate(letters)) if(i %% 2==0) paste(i, j))

set.seed(123)
rand_sequence = runif(20)
# gives only locally increasing values
to_vec(for(`i, j` in lag_list(rand_sequence)) if(j>i) j)


## ------------------------------------------------------------------------
data(iris)
# scale numeric variables
res = alter(for(i in iris) if(is.numeric(i)) scale(i))
str(res)

# convert factors to characters
res = alter(for(i in iris) if(is.factor(i)) as.character(i))
str(res)

# drop factors
res = alter(for(i in iris) if(is.factor(i)) exclude())
str(res)

# 'data' argument example
# specify which columns to map with a numeric vector of positions:
res = alter(
    for(`i, value` in numerate(mtcars)) if(i %in% c(1, 4, 5)) as.character(value),
    data = mtcars
)
str(res)

# or with a vector of names:
res = alter(
    for(`name, value` in mark(mtcars)) if(name %in% c("cyl", "am")) as.character(value),
    data = mtcars
)
str(res)



R/to_list.R
SPECIAL_FUNCTIONS = expression(
    to_list,
    to_vec,
    alter ,
    to_df ,
    to_dfr,
    to_dfc,
    comprehenr::to_list,
    comprehenr::to_vec ,
    comprehenr::alter  ,
    comprehenr::to_df  ,
    comprehenr::to_dfr ,
    comprehenr::to_dfc
)


#' List comprehensions for R
#'
#' - `to_list` converts usual R loops expressions to list producers.
#' Expression should be started with `for`, `while` or
#' `repeat`. You can iterate over multiple lists if you provide several
#' loop variables in backticks. See examples.
#' - `to_vec` is the same as 'to_list' but return vector. See examples.
#' - `to_df` is the same as 'to_list' but return data.frame. All elements of
#' resulted list will be converted to data.frame and combined via `rbind`.
#' - `alter` returns the same type as its argument but with modified
#' elements. It is useful for altering existing data.frames or lists. See
#' examples.
#' - `exclude` is an auxiliary function for dropping elements in
#' `alter`. There are no arguments for this function.
#'
#' @param expr expression which starts with `for`, `while` or `repeat`.
#' @param recursive	logical. Should unlisting be applied to list components of result? See [unlist][base::unlist] for details.
#' @param use.names logical. Should names be preserved? See [unlist][base::unlist] for details.
#' @param fill logical. TRUE by default. Should we combine data.frames with different names in the `to_df`?
#' @param data data.frame/list/vector which we want to alter
#' @return list for `to_list` and vector for `to_vec`
#' @export
#'
#' @examples
#' # rather useless expression - squares of even numbers
#' to_list(for(i in 1:10) if(i %% 2==0) i*i)
#'
#' # Pythagorean triples
#' to_list(for (x in 1:30) for (y in x:30) for (z in y:30) if (x^2 + y^2 == z^2) c(x, y, z))
#'
#' colours = c("red", "green", "yellow", "blue")
#' things = c("house", "car", "tree")
#' to_vec(for(x in colours) for(y in things) paste(x, y))
#'
#' # prime numbers
#' noprimes = to_vec(for (i in 2:7) for (j in seq(i*2, 99, i)) j)
#' primes = to_vec(for (x in 2:99) if(!x %in% noprimes) x)
#' primes
#'
#' # iteration over multiple lists
#' to_vec(for(`i, j` in numerate(letters)) if(i %% 2==0) paste(i, j))
#'
#' set.seed(123)
#' rand_sequence = runif(20)
#' # gives only locally increasing values
#' to_vec(for(`i, j` in lag_list(rand_sequence)) if(j>i) j)
#'
#' # to_df
#' to_df(for(`name, x` in mark(mtcars)) list(mean = mean(x), sd = sd(x), var = name))
#'
#' # 'alter' examples
#' data(iris)
#' # scale numeric variables
#' res = alter(for(i in iris) if(is.numeric(i)) scale(i))
#' str(res)
#'
#' # convert factors to characters
#' res = alter(for(i in iris) if(is.factor(i)) as.character(i))
#' str(res)
#'
#' # exclude factors from data.frame
#' res = alter(for(i in iris) if(is.factor(i)) exclude())
#' str(res)
#'
#' # 'data' argument example
#' # specify which columns to map with a numeric vector of positions:
#' res = alter(
#'     for(`i, value` in numerate(mtcars)) if(i %in% c(1, 4, 5)) as.character(value),
#'     data = mtcars
#' )
#' str(res)
#'
#' # or with a vector of names:
#' res = alter(
#'     for(`name, value` in mark(mtcars)) if(name %in% c("cyl", "am")) as.character(value),
#'     data = mtcars
#' )
#' str(res)
to_list = function(expr){
    expr = substitute(expr)
    is_loop(expr) ||
        stop(paste("argument should be expression with 'for', 'while' or 'repeat' but we have: ", deparse(expr, width.cutoff = 500)[1]))


    expr = expand_loop_variables(expr)
    expr = add_assignment_to_final_loops(expr)

    expr = substitute(local({
        .___res <- list()
        .___counter <- 0  # initial list length
        expr
        .___res
    }))

    eval.parent(expr)

}

#' @export
#' @rdname to_list
to_vec = function(expr, recursive = TRUE, use.names = FALSE){
    res= eval.parent(substitute(comprehenr::to_list(expr)))
    unlist(res, recursive = recursive, use.names = use.names)
}


add_assignment_to_final_loops = function(expr, result_exists = FALSE){
    if(is_loop(expr)){
        if(has_loop_inside(expr[-1])){
            expr[-1] = as.call(lapply(as.list(expr)[-1], add_assignment_to_final_loops, result_exists = result_exists))
        } else {
            expr = add_assignment_to_loop(expr, result_exists = result_exists)
        }
    } else {
        if(has_loop_inside(expr)){
            expr = as.call(lapply(as.list(expr), add_assignment_to_final_loops, result_exists = result_exists))
        }
    }
    expr
}

expand_loop_variables = function(expr){
    if(!is.call(expr)) return(expr)
    if(is_for_loop(expr)){
            expr = add_expansion_to_loop(expr)
            expr[-1] = as.call(lapply(as.list(expr)[-1], expand_loop_variables))
    } else {
        expr = as.call(lapply(as.list(expr), expand_loop_variables))
    }
    expr
}

add_expansion_to_loop = function(expr){

    loop_variable = expr[[2]]
    if(is.symbol(loop_variable) && grepl(",", as.character(loop_variable))){
        all_loop_variables = trimws(unlist(strsplit(as.character(loop_variable), split = ",")))
        expansion_expr = lapply(seq_along(all_loop_variables), function(item_num){
            curr_item = as.name(all_loop_variables[item_num])
            bquote(.(curr_item) <- .(loop_variable)[[.(item_num)]])
        })
        expansion_expr = as.call(c(list(quote(`{`)), expansion_expr))
        last_item = length(expr)
        expr[[last_item]] = bquote({
            .(expansion_expr)
            .(expr[[last_item]])
        })
    }
    expr
}

is_loop = function(expr){
    is.call(expr) || return(FALSE)
    first_item = expr[[1]]
    identical(first_item, quote(`for`)) ||identical(first_item, quote(`while`)) ||identical(first_item, quote(`repeat`))
}

is_for_loop = function(expr){
    is.call(expr) || return(FALSE)
    first_item = expr[[1]]
    identical(first_item, quote(`for`))
}

has_loop_inside = function(expr){
    is.call(expr) || return(FALSE)
    !is_loop(expr) || return(TRUE)
    first_elem = expr[[1]]
    for(item in SPECIAL_FUNCTIONS){
        if(identical(item, first_elem)) return(FALSE)
    }
    any(
        unlist(
            lapply(as.list(expr), has_loop_inside),
            recursive = TRUE,
            use.names = FALSE
        )
    )
}


add_assignment_to_loop = function(expr, result_exists = FALSE){
    last_item = length(expr)
    if(result_exists){
        expr[[last_item]] = bquote({
            .___curr = {.(expr[[last_item]])}
            if(inherits(.___curr, "exclude")){
                .___res[[.___counter + 1]] = NULL
            } else {
                .___counter = .___counter + 1
                if(!is.null(.___curr)){
                    .___res[[.___counter]] = .___curr
                }
            }

        })
    } else {
        expr[[last_item]] = bquote({

            .___curr = {.(expr[[last_item]])}
            if(!is.null(.___curr) && !inherits(.___curr, "exclude")){
                .___counter = .___counter + 1
                .___res[[.___counter]] = .___curr
            }

        })
    }
    expr
}

#' @rdname to_list
#' @export
alter = function(expr, data = NULL){
    expr = substitute(expr)
    is_loop(expr) ||
        stop(paste("argument should be expression with 'for', 'while' or 'repeat' but we have: ", deparse(expr, width.cutoff = 500)[1]))

    if(is.null(data)) {
        data = expr[[3]]
        if(is.call(data)){
            first = as.character(data[[1]])
            if(first %in% c("numerate", "enumerate", "mark", "lag_list")){
                data = data[[2]]
            }
        }

    }
    expr = expand_loop_variables(expr)
    expr = add_assignment_to_final_loops(expr, result_exists = TRUE)
    expr = substitute(local({
        .___res <- data
        .___counter <- 0  # initial list length
        expr
        .___res
    }))

    eval.parent(expr)

}


#' @rdname to_list
#' @export
to_df = function(expr, fill = TRUE){
    res = eval.parent(
        substitute(
            comprehenr::to_list(expr)
        )
    )
    for(i in seq_along(res)){
        if(!is.null(res[[i]]) && !is.data.frame(res[[i]])){
            res[[i]] = as.data.frame(as.list(res[[i]]))
        }
    }
    #
    if(fill){
        all_names = unique(unlist(lapply(res, names), use.names = FALSE))
        for(i in seq_along(res)){
            new_names = setdiff(all_names, names(res[[i]]))
            if(length(new_names)>0){
                res[[i]][,new_names] = NA
            }
        }
    }

    do.call(rbind, res)
}

# @rdname to_list
# @export
# to_dfr = to_df

# @rdname to_list
# @export
# to_dfc = function(expr = NULL){
#     res = eval.parent(
#         substitute(
#             comprehenr::to_list(expr)
#         )
#     )
#     as.data.frame(res)
# }


#' @rdname to_list
#' @export
exclude = function(){
    res = numeric(0)
    class(res) = "exclude"
    res
}
